Add `bitcoind_rpc` chain source module.

evanlinjin opened the pull request:
### Description

This PR builds on top of #1034 and adds the `bitcoind_rpc` chain-src module and example.

### Notes to the reviewers

Don't merge this until #1034 is in!

### Changelog notice

* Add `bitcoind_rpc` chain-source module.
* Add `example_bitcoind_rpc` example module.
* Add `AnchorFromBlockPosition` trait which are for anchors that can be constructed from a given block, height and position in block.
* Add helper methods to `IndexedTxGraph` and `TxGraph` for batch operations and applying blocks directly.
* Add helper methods to `CheckPoint` for easier construction from a block `Header`.

### Checklists

* [x] Add test: we should detect when an initially-confirmed transaction is "unconfirmed" during a reorg.
* [x] Improve `example_bitcoind_rpc`: add `live` command.
* [x] Improve docs.
* [x] Reintroduce `CheckPoint`.

#### All Submissions:

* [x] I've signed all my commits
* [x] I followed the [contribution guidelines](https://github.com/bitcoindevkit/bdk/blob/master/CONTRIBUTING.md)
* [x] I ran `cargo fmt` and `cargo clippy` before committing

#### New Features:

* [x] I've added tests for the new feature
* [x] I've added docs for the new feature

notmandatory commented on crates/bitcoind_rpc/Cargo.toml:
```diff
@@ -0,0 +1,14 @@
+[dependencies]
+bdk_chain = { path = "../chain", version = "0.5.0", features = ["serde", "miniscript"] }
+bitcoincore-rpc = { version = "0.17.0" }
+anyhow = { version = "1" }
```
Do we need to add `anyhow` ? None of the other crates use it, so would like to not have to add it for this one.

evanlinjin replied:
Woops, I think that was only meant as a dev dependency! Thanks for pointing this out.

LLFourn commented on crates/bitcoind_rpc/src/lib.rs:
```diff
@@ -0,0 +1,476 @@
+//! This crate is used for updating [`bdk_chain`] structures with data from the `bitcoind` RPC
```
But not the rpc wallet API :warning: :warning:

LLFourn commented on example-crates/example_rpc/src/main.rs:
```diff
@@ -0,0 +1,308 @@
+                        let anchor_map = bdk_bitcoind_rpc::confirmation_time_anchor;
+                        item.into_tx_graph_update(tx_filter, anchor_map)
+                    };
+                    indexed_additions.append(graph.apply_update(graph_update));
```
This seems to be more complicated than it needs to be. Why can't I just pass off a bunch of transactions whether they be in a block or just a chunk of mempool transactions to the indexed tx graph and have it tell me the changeset after applying them. Passing off the `indexer` to the bitcoin_rpc logic seems like a hack. bitcoind_rpc shouldn't know about this type. It's a totally separate concern.

evanlinjin replied:
This is addressed in 26209de743c4cd134fcd53a1db2b7c121b27230c

LLFourn commented on crates/bitcoind_rpc/src/lib.rs:
```diff
@@ -0,0 +1,476 @@
+#[derive(Debug, Clone)]
+pub struct EmittedBlock {
+    /// The checkpoint constructed from the block's height/hash and connected to the previous block.
+    pub cp: CheckPoint,
```
Is this needed. Won't the logic receiving this block be able to figure out where to connect it. `Block` has the previous block hash in it which should have already been emitted if the caller didn't know about it already.

evanlinjin replied:
I think it makes our API a bit more flexible.

For example:
* The caller pulled data from esplora to height 100.
* The caller waits a while.
* The caller creates wallet with birthday at height 200.

If we create a `local_chain::Update` at height 200 with checkpoints derived from `Block` only, we can't connect to the chain at height 100.

LLFourn replied:
> I think it makes our API a bit more flexible.
> 
> For example:
> 
>     * The caller pulled data from esplora to height 100.
> 
>     * The caller waits a while.
> 
>     * The caller creates wallet with birthday at height 200.
> 
> 
> If we create a `local_chain::Update` at height 200 with checkpoints derived from `Block` only, we can't connect to the chain at height 100.

In this case you would pass in the block at `100` as `last_cp` to the emitter and then it would emit every block from 100 to 200 in order which means it would connect to `100`. Am I missing something?

evanlinjin replied:
What if we want to sync from 200 without removing old blocks?

What if the caller only has the Genesis block and wallet birthday is at height 805016

LLFourn replied:
These are certainly questions. What are the answers? :sweat_smile:

evanlinjin replied:
@LLFourn I'm very confused what you mean by this comment haha.

LLFourn replied:
I am trying to find out the answer to my question and the questions you posed in response to my question.

LLFourn commented on crates/bitcoind_rpc/src/lib.rs:
```diff
@@ -0,0 +1,476 @@
+#[derive(Debug, Clone)]
+pub struct EmittedMempool {
+    /// The checkpoint of the last-seen tip.
+    pub cp: CheckPoint,
```
please explain in the docs how I am meant to use this information.

evanlinjin replied:
I removed `EmittedMempool::cp`. It's not useful.

LLFourn commented on crates/bitcoind_rpc/src/lib.rs:
```diff
@@ -0,0 +1,476 @@
+    /// The checkpoint of the last-seen tip.
+    pub cp: CheckPoint,
+    /// Subset of mempool transactions.
+    pub txs: Vec<(Transaction, u64)>,
```
document what the `u64` is for.

LLFourn commented on example-crates/example_rpc/src/main.rs:
```diff
@@ -0,0 +1,308 @@
+                        }
+                    }
+                }
+                unreachable!()
```
This `unreachable` thing biases me against providing this iterator API. It ended up conceptually weird to me. IWouldn't it be clearer just to emit blocks and process them in a `loop`.

evanlinjin replied:
The way we are using it here is a bit weird. We have a flag for both "live" mode and "sync-once" mode. In my opinion, the iterator will be useful in a "sync-once" mode. The "unreachable" thing is not a necessity or a direct result of using the iterator.

LLFourn commented on example-crates/example_rpc/src/main.rs:
```diff
@@ -0,0 +1,308 @@
+    };
+
+    match rpc_cmd {
+        RpcCommands::Scan {
```
Maybe just `sync` instead of `scan` since we use that for scanning addresses. This is rather syncing local state with remote state like the sync commands.

LLFourn reviewed (commented):
I've made some comments. It looks awesome. I think this will be a "flagship" example on how to use bdk so I think we should spend the time to get it right. The block data processing pipeline is too complicated for what the problem is. I think `IndexedTxGraph` needs to be able to process block data and filter out what's relevant itself. This is needed for CBF block processing as well so calling out to weird helper methods in the bitcoind_rpc crate seems counter productive.

notmandatory commented:
@evanlinjin this looks like a really powerful way to sync with a core node. I read through the code and I think I understand the state-model approach within the `Emitter::emit_block`. Have you and @LLFourn chatted yet about his ideas to simplify it somehow? From an outside user the APIs look easy enough to use but if there are opportunities to simplify the internals it's worth discussing.

You look to be covering the main test scenarios, 1. sync blocks before and after reorg, 2. sync txs in mempool and after being included in a block.  Is it worth adding a test to verify confirmed tx are unconfirmed after a reorg?

notmandatory commented:
Looks like we have another MSRV issue. Need to add these:

in `cont_integration.yml`
```
          cargo update -p log --precise "0.4.18"
          cargo update -p tempfile --precise "3.6.0"
          cargo update -p minreq --precise "2.8.1"              <======= add this line to fix
          cargo update -p rustls:0.21.6 --precise "0.21.1"
          cargo update -p tokio:1.32.0 --precise "1.29.1"
          cargo update -p flate2:1.0.27 --precise "1.0.26"
          cargo update -p reqwest --precise "0.11.18"
          cargo update -p h2 --precise "0.3.20"
          cargo update -p rustls-webpki --precise "0.100.1"
```

in `README.md` 
```
# log 0.4.19 has MSRV 1.60.0+
cargo update -p log --precise "0.4.18"
# tempfile 3.7.0 has MSRV 1.63.0+
cargo update -p tempfile --precise "3.6.0"
# minreq 2.9.0 prevents pinning rustls to 0.21.1        <====== add corresponding docs
cargo update -p minreq --precise "2.8.1"
# rustls 0.21.2 has MSRV 1.60.0+
cargo update -p rustls:0.21.6 --precise "0.21.1"
# tokio 1.30 has MSRV 1.63.0+
cargo update -p tokio:1.32.0 --precise "1.29.1"
# flate2 1.0.27 has MSRV 1.63.0+
cargo update -p flate2:1.0.27 --precise "1.0.26"
# reqwest 0.11.19 has MSRV 1.63.0+
cargo update -p reqwest --precise "0.11.18"
# h2 0.3.21 has MSRV 1.63.0+
cargo update -p h2 --precise "0.3.20"
# rustls-webpki 0.100.2 has MSRV 1.60.0+.                         <======== small update
cargo update -p rustls-webpki --precise "0.100.1"
```

notmandatory commented:
@vladimirfomene is working on an rpc wallet example and reorg confirmed transaction tests.

RCasatta commented on crates/bitcoind_rpc/tests/test_emitter.rs:
```diff
@@ -0,0 +1,300 @@
+
+impl TestEnv {
+    fn new() -> anyhow::Result<Self> {
+        let daemon = match std::env::var_os("TEST_BITCOIND") {
+            Some(bitcoind_path) => bitcoind::BitcoinD::new(bitcoind_path),
+            None => bitcoind::BitcoinD::from_downloaded(),
+        }?;
+        let client = bitcoincore_rpc::Client::new(
+            &daemon.rpc_url(),
+            bitcoincore_rpc::Auth::CookieFile(daemon.params.cookie_file.clone()),
+        )?;
```
This works ok but I was wondering if there are reasons not to use
```
 // use BITCOIND_EXE instead of TEST_BITCOIND, see https://docs.rs/bitcoind/0.33.0/bitcoind/fn.exe_path.html
let daemon = bitcoind::BitcoinD::new(bitcoind::exe_path()?)?; 
```
and then using the inner `daemon.client` instead of creating a new one

evanlinjin replied:
I think that will work with the latest version of `bitcoind`. However, we were using the old version before, where there was a custom package name for the `bitcoincore-rpc` dependency. Will change soon!

evanlinjin replied:
Updated in c87b1b08bbfe63c22961093572b16828d9c355fe

---

Edit: I had to revert c87b1b08bbfe63c22961093572b16828d9c355fe because latest version of `bitcoind` is breaking MSRV.

notmandatory replied:
I was able to use to the latest version of `electrsd` (which depends on latest 0.33 `bitcoind`) without breaking MSRV with this commit: https://github.com/bitcoindevkit/bdk/pull/1110/commits/7be6f86343e30c9e07bcfd52342a203ea7556026

So I think we should be able to keep latest `bitcoind`.

evanlinjin replied:
I have upgraded to the latest `bitcoind` crate. However, I think it is best to keep the implementation as is, just in case `bitcoincore_rpc` versions diverges between `bitcoind` and our dependencies?

evanlinjin commented:
@LLFourn suggested that we try "decouple" the `bdk_bitcoind_rpc` API from `bdk_chain` structures. The reason for this is as follows:

* `bdk_chain` structures should be directly updateable by inputting raw bitcoin data such as `Block`s, `Transaction`s, etc. We shouldn't need to *transform* the data into `bdk_chain` types before applying it to `TxGraph`, `LocalChain`, etc.

* The will allow us to update the `bdk_chain` and `bdk_bitcoind_rpc` crates orthogonal to one another.

* This will make `bdk_bitcoind_rpc` better when used with non-`bdk_chain` projects. This will reduce a dependency.

LLFourn commented on example-crates/example_rpc/src/main.rs:
```diff
@@ -0,0 +1,346 @@
+                    Emission::Mempool(txs) => {
+                        txs.iter().map(|tx| (&tx.tx, None, Some(tx.time))).collect()
+                    }
+                    Emission::Block(b) => {
```
Can we have `apply_block(&block, height)` on `TxGraph` and `IndexedTxGraph`? I don't think you want users doing this manually when it should be that simple. The anchor type of `A` for `TxGraph` should be constrained so that it can be extracted from a block header and the position of the transaction in the block. Probably need a new trait for this like `FromBlockPosition`. `FromBlockPosition` can take `(&block, position)`

We could take this opportunity to have a different anchor for this one which includes block position.

`IndexedTxGraph` should also have `apply_block_relevant` or something like that which only adds the relevant transactions which you'd use here.

For mempool dumping I'd add `batch_insert_unconfirmed` and `batch_insert_relevant_unconfirmed`. These can take a iter of `(tx, last_seen)`.

evanlinjin replied:
@LLFourn I don't think having `apply_block` on `TxGraph` makes sense, since we have no `Indexer`. I'll add it to `IndexedTxGraph` though.

LLFourn replied:
I'm not sure what you meant here. Do you mean you think we should only add `apply_block_relevant` and not add `apply_block` to either of them?

I would add `apply_block` too. You can use the `TxGraph` for general analysis. Maybe I want to know what % of new outputs are spent within the next 10 blocks etc. You don't need an indexer to make `TxGraph` useful.

evanlinjin replied:
Woops, I assumed that you meant `apply_block` would filter for relevant txs (as that was what the original codebase did) and didn't finish reading your comment.

evanlinjin replied:
introduced here: 09bdffab2fd08290157c0f0e31058dcaefbbe308

LLFourn commented on example-crates/example_rpc/src/main.rs:
```diff
@@ -0,0 +1,346 @@
+        lookahead: u32,
+        /// Whether to be live!
+        #[clap(long, default_value = "false")]
+        live: bool,
```
Sorry to request this late. Can we have two command: `sync` and `live`. The `sync` command handler should just have the minimal logic to apply all the latest blocks and the mempool and then exit (nice to print how many blocks and mempool txs were processed). The live command should do what the current code does. This is so a developers can more easily see how to interact with the crate in the two cases.

LLFourn commented on example-crates/example_rpc/src/main.rs:
```diff
@@ -0,0 +1,346 @@
+                &chain,
+                &keymap,
+                args.network,
+                |_| Err(anyhow::anyhow!("use `tx` instead")),
```
Why?

evanlinjin replied:
We cannot construct the `Client` before knowing the client's url and params, and they are defined in the `Commands::ChainSpecific` enum's variants. We cannot modify `Commands::Send`.

LLFourn replied:
Just for ref I made a PR to this branch fixing this: https://github.com/evanlinjin/bdk/pull/7

This PR can be merged now or I can re-do it to master after this is merged.

evanlinjin replied:
Rebased and added to this PR as 2db28e0376e8f7b195d78e23cd5283e7a5c21a45

LLFourn commented on crates/bitcoind_rpc/src/lib.rs:
```diff
@@ -0,0 +1,321 @@
+/// Refer to [module-level documentation] for more.
+///
+/// [module-level documentation]: crate
+pub enum Emission<B> {
```
Why do these different data requests (mempool, next block) needed to be handled in the same "stream" of events. Looking at the caller code it doesn't seem to simplify things, they are always matched on and handled differently (this could be because of the difference in vision I have kept pushing you into a corner where this was the case).

Furthermore, A user is likely to want to check for the next block vs dump the mempool at different times. E.g. dump the mempool every 30s but check for next block every 5s (actually it would be nice if we did this in our example). I strongly believe the caller should make a conscious decision about when to do this. It is not a natural decision to me to dump the mempool after you've finished looking for new blocks (and then continuing to dump it at every second call to `next` which feels even weirder).

evanlinjin replied:
If we scan the mempool for relevant txs before scanning the latest tip, we may end up excluding mempool transactions that send from our UTXO pool (as we need the prev txs to determine relevancy).

This may be problematic if we intend to have non-duplicate tx dumps from the mempool (as suggested in https://github.com/bitcoindevkit/bdk/pull/1041#discussion_r1315406687). We will end up excluding an actually-relevant mempool transaction which sends from our wallet.

If this tx takes a while to confirm, we can be fully synced up to the tip and not see it in the wallet until it is confirmed. The wallet may accidentally rbf the tx if the tx stays unconfirmed.

I think it is good policy to scan the mempool only after scanning the latest tip.

LLFourn commented on crates/bitcoind_rpc/src/lib.rs:
```diff
@@ -0,0 +1,321 @@
+
+    /// Emit all mempool transactions.
+    pub fn mempool(&self) -> impl Iterator<Item = Result<MempoolTx, bitcoincore_rpc::Error>> + '_ {
+        let res = self.client.get_raw_mempool();
```
side question: there a way of custom dumping the mempool so you only get transactions that entered after a certain time?

notmandatory replied:
I don't see anything in the RPC options for `get_raw_mempool` to filter what it returns. Probably the only way to make this more efficient is to plug into a `zmq` topic like `rawtx` and get block and mempool tx updates that way. But figuring out if/how to use `zmq` should be a post 1.0 thing to do.

https://bitcoincore.org/en/doc/25.0.0/rpc/blockchain/getrawmempool/

evanlinjin replied:
We can use `get_raw_mempool` to only gets txids. Then, we can use the `get_mempool_entry` on an individual txid to get the time that the tx entered the mempool.

LLFourn commented on example-crates/example_rpc/Cargo.toml:
```diff
@@ -0,0 +1,12 @@
+[package]
+name = "example_rpc"
```
please rename directory and package `example_bitcoind_rpc`. Actually it should probably be `example_bitcoind_rpc_block_polling`.

notmandatory commented:
If you do a rebase it should fix the MSRV errors. The only job not passing right now is the audit job ☹️, but I have an indirect fix for that in https://github.com/RCasatta/bitcoind/pull/141.

evanlinjin commented on crates/bitcoind_rpc/src/lib.rs:
```diff
@@ -0,0 +1,306 @@
+    ///
+    /// This avoids re-emitting transactions (where viable). We can do this if all blocks
+    /// containing ancestor transactions are already emitted.
+    pub fn mempool(&mut self) -> Result<Vec<MempoolTx>, bitcoincore_rpc::Error> {
```
We should have the following tests checking the behavior when `Emitter::mempool` is called multiple times.

- [x] The receiver (`bdk_chain` structures) is synced to the chain tip, and there is txs in the mempool. When we call `Emitter::mempool` multiple times, mempool txs should not be re-emitted.

- [x] In the above scenario, if the network's chain extends with empty blocks (and our receiver stays up to sync), we expect that the txs do not get re-emitted.

- [x] The network is at height `h`. The receiver is synced to height `h-1`. When a mempool tx `A` is introduced at height `h`, we should always re-emit it.

- [x] In the above scenario, if the receiver syncs up to `h`, we should still re-emit (right after). However, subsequent calls to `Emitter::mempool` should not re-emit.

notmandatory replied:
I added a test for the 3rd checkbox and a CI fix in #1140, if it looks OK feel free to copy/paste or cherry-pick, which ever is easier.

evanlinjin replied:
@notmandatory Thank you for the PR! Unfortunately we duplicated work, and I think my solution covers a broader set of situations, so will stick with that.

notmandatory replied:
No problem! closed my PR.

LLFourn commented on crates/bitcoind_rpc/src/lib.rs:
```diff
@@ -0,0 +1,306 @@
+                    }
+
+                    // Avoid emitting transactions that are already emitted if we can guarantee
+                    // blocks containing ancestors are already emitted. We only check block height
```
This comment doesn't make sense to me. I don't see anything where you guarantee blocks containing ancestors are already emitted (but also don't understand why that's relevant).

evanlinjin replied:
The `Indexer` may think a transaction is irrelevant for the time being (since it does not have the prev outpoints that it owns). We should keep re-emitting such a transaction.

A bitcoind's tx entry gives us the block height in which the tx is introduced to the mempool. If our `Emitter`'s last-emitted-block-state meets or surpasses this block height, we can guarantee that we have scanned or ancestors (except in case of reorg).

For reorg, we force-re-emit everything.

notmandatory commented on crates/bitcoind_rpc/src/lib.rs:
```diff
@@ -0,0 +1,322 @@
+//!
+//! [`Emitter`] is the main structure which sources blockchain data from [`bitcoincore_rpc::Client`].
+//!
+//! To only get block updates (exlude mempool transactions), the caller can use
```
nit: typo
```suggestion
//! To only get block updates (exclude mempool transactions), the caller can use
```

notmandatory reviewed (approved):
ACK 409033a8cf75e8216dc8611d7708e8703e36fcc2

A few little nits need fixing, but overall looks like a powerful way to sync via bitcoind rpc. But since it's so new I recommend we put a little "experimental" disclaimer in the release notes until it's gotten more user real-world testing.

notmandatory commented:
I created a little test for the "no agreement point" situation. Feel free to cherry pick or copy/pasta if you think it's useful.

https://github.com/bitcoindevkit/bdk/pull/1143/commits/e728c53c0d3406ea66aa9fba59953f7188db7c1a

evanlinjin committed:
feat(example_cli): allow chain specific args in examples

So you can pass in the esplora/electrum/bitcoind_rpc server details in
the example.

Co-authored-by: LLFourn <lloyd.fourn@gmail.com>

evanlinjin committed:
feat(chain): add `AnchorFromBlockPosition` trait

This is useful for block-by-block chain sources. We can determine the
tx's anchor based on the block, block height and tx position in the
block.

evanlinjin committed:
chain: add helper methods on `CheckPoint`

* `CheckPoint::from_header` allows us to construct a checkpoint from
  block header.
* `CheckPoint::into_update` transforms the cp into a
  `local_chain::Update`.

evanlinjin committed:
chain: add batch-insert methods for `IndexedTxGraph`

evanlinjin committed:
feat(bitcoind_rpc): introduce `bitcoind_rpc` crate

evanlinjin committed:
test(bitcoind_rpc): initial tests for `Emitter`

evanlinjin committed:
test(bitcoind_rpc): add no_agreement_point test

Co-authored-by: Steve Myers <steve@notmandatory.org>

evanlinjin committed:
feat(example_bitcoind_rpc_polling): add example for RPC polling

evanlinjin commented:
The `Emitter::mempool` method is taking ~20seconds (today, with a busy mempool). This is because we need call `getrawtransaction` for every mempool entry that requires re-emission.

Possible ways to optimize in the future:

1. Cache transactions in the emitter.
2. Use multiple threads to call `getrawtransaction`.
3. A combination of the above.

This issue should not exist if we sync to tip before emitting from mempool (need to check).

LLFourn commented on example-crates/example_bitcoind_rpc_polling/src/main.rs:
```diff
@@ -0,0 +1,366 @@
+
+            let mut chain = chain.lock().unwrap();
+            let mut graph = graph.lock().unwrap();
+            let mut db = db.lock().unwrap();
```
:x: don't keep locks open over IO.

evanlinjin replied:
But they aren't being used anywhere else so we'll we wasting resources locking/unlocking?

LLFourn replied:
But if they were being used somewhere else the whole application would be stuck. We're in an example. The point of it is to show people (and ourselves) how to use the API in a thread safe context so that other parts of the application can access the these components while a sync is going on.

evanlinjin replied:
I'll change it, but it bugs me. I'm wondering whether we need to show developers how to use mutexes in our examples. For me, it would seem like common sense to know how to wrap any structure with mutexes/locks.

LLFourn commented on crates/chain/src/indexed_tx_graph.rs:
```diff
@@ -145,27 +141,149 @@ where
+    /// To filter out irrelevant transactions, use [`batch_insert_relevant_unconfirmed`] instead.
+    ///
+    /// [`batch_insert_relevant_unconfirmed`]: IndexedTxGraph::batch_insert_relevant_unconfirmed
+    pub fn batch_insert_unconfirmed<'t>(
```
Please add the same method to `TxGraph`.

LLFourn commented on crates/chain/src/indexed_tx_graph.rs:
```diff
@@ -145,27 +141,149 @@ where
+    /// Batch insert unconfirmed transactions.
+    ///
+    /// Items of `txs` are tuples containing the transaction and an optional *last seen* timestamp.
+    /// The *last seen* communicates when the transaction is last seen in the mempool which is used
```
What's the motivation for making `last_seen` optional. I think it should be mandatory.

LLFourn commented on crates/chain/src/indexed_tx_graph.rs:
```diff
@@ -145,27 +141,149 @@ where
+///
+/// [`batch_insert_relevant`]: IndexedTxGraph::batch_insert_relevant
+/// [`batch_insert`]: IndexedTxGraph::batch_insert
+pub type InsertTxItem<'t, A> = (&'t Transaction, A, Option<u64>);
```
Not sure about introducing this concept. Is it so bad to ask the user to insert anchors/last_seen using their separate calls?

[EDIT] I just realized they are unfortunately not separate calls in indexed tx graph and made a comment elsewhere about fixing that.

evanlinjin replied:
~~What if we want to update with a mixed list of unconfirmed/confirmed transactions? I would prefer it if we just had `batch_insert` and `batch_insert_relevant` which takes in `InsertTxItem` (we can then get rid of the other methods). I think this is better because having too many methods that does similar but slightly different things can be very confusing.~~

**EDIT:** After thinking about it, we only need batch methods for unconfirmed anyway since it requires additional logic.

LLFourn commented on crates/chain/src/indexed_tx_graph.rs:
```diff
@@ -145,27 +141,149 @@ where
+    /// [`batch_insert_relevant`] instead.
+    ///
+    /// [`batch_insert_relevant`]: IndexedTxGraph::batch_insert_relevant
+    pub fn batch_insert<'t>(
```
Don't think this method is worth it. Doesn't do anything beyond calling `insert_tx` in a loop and appending changesets (it's surprising that it's not implemented like this).

evanlinjin replied:
Do you mind expanding on what you mean by what's not implemented like this?

evanlinjin replied:
I think I understand what you mean now. This method is now removed.

LLFourn commented on example-crates/example_cli/src/lib.rs:
```diff
@@ -664,45 +666,213 @@ where
+                    println!("Broadcasted Tx : {}", transaction.txid());
+
+                    let keychain_changeset =
+                        graph.lock().unwrap().insert_tx(&transaction, None, None);
```
Why does indexed tx graph `insert_tx` take three arguments but tx graph only take one. Eeek. We should make this the same. I like `TxGraph`'s API better FWIW.

evanlinjin replied:
I guess the only benefit of having `insert_tx` also insert anchors and last_seen is we can reduce the number of times we calculate the txid :sweat_smile:.

I agree that the API between `TxGraph` and `IndexedTxGraph` should be relatively uniform. So `IndexedTxGraph` should have separate methods for inserting tx, anchor and last_seen.

However, I think this should be tackled in a separate issue.

LLFourn replied:
> I guess the only benefit of having `insert_tx` also insert anchors and last_seen is we can reduce the number of times we calculate the txid 😅.

How does it?

> However, I think this should be tackled in a separate issue.

Why? Does it need further discussion? I'd just do this now and forget about it.

evanlinjin replied:
Done.

LLFourn commented on crates/chain/src/local_chain.rs:
```diff
@@ -39,6 +40,39 @@ impl CheckPoint {
+        let hash = header.block_hash();
+        let this_block_id = BlockId { height, hash };
+
+        if header.prev_blockhash == BlockHash::all_zeros() {
```
isn't `if height == 0` strictly more correct here? This avoids the underflow on line 57 also. `match height.checked_sub(1)` might be the way to go.

LLFourn commented on crates/chain/src/local_chain.rs:
```diff
@@ -39,6 +40,39 @@ impl CheckPoint {
+        let prev_block_id = BlockId { height, hash };
+
+        CheckPoint::new(prev_block_id)
+            .extend(core::iter::once(this_block_id))
```
don't we have a method like `push` for single things?

LLFourn commented on example-crates/example_bitcoind_rpc_polling/src/main.rs:
```diff
@@ -0,0 +1,366 @@
+            });
+
+            let rpc_client = rpc_args.new_client()?;
+            let mut emitter = Emitter::new(&rpc_client, start_height);
```
Hmm iirc this used to be a checkpoint instead of `start_height` so we didn't have this `ASSUME_FINAL_DEPTH` thing. That was much better but I see that it needed to be sacrificed to remove the depend on `bdk_chain`. Could we make it an `Iterator` of `(height, hash)` and make an easy way to create one of those from a `CheckPoint` and use the same clever logic we had before?

Either that or we could have a `bdk_core` crate that just defines some types like `BlockId`, `Checkpoint` that we can hopefully not make breaking changes to (at least no where near as many as we'll make to `bdk_chain`).

evanlinjin replied:
I like the `bdk_core` crate idea. If we decide to go down this approach, let's make these changes a separate issue?

**EDIT:** I would prefer it if this version of `bitcoind_rpc` still depended on `bdk_chain` for initializing `Emitter` with `CheckPoint`. Then we introduce the `bdk_core` crate in a later version?

LLFourn replied:
Sounds like a plan. The most important thing is the code has been written not to depend on details of `bdk_chain` at least and we can switch it to `bdk_core` later.

evanlinjin replied:
Made an issue: https://github.com/bitcoindevkit/bdk/issues/1155

LLFourn commented on example-crates/example_bitcoind_rpc_polling/src/main.rs:
```diff
@@ -0,0 +1,366 @@
+            while let Some((height, block)) = emitter.next_block()? {
+                let chain_update =
+                    CheckPoint::from_header(&block.header, height).into_update(false);
+                let chain_changeset = chain.apply_update(chain_update)?;
```
Can we guarantee that emitter won't give us out of order blocks here? If so I would `expect` this. If not maybe we can figure out how we can give this guarantee?

evanlinjin replied:
~We can guarantee that emitter won't give us out of order blocks here.~

**EDIT:** Actually there is one scenario where the emitted block won't connect. It's when there is a reorg that happens lower than our start_height. We'll need to make use of `CheckPoint`.

LLFourn replied:
Yes we need to pass in the iterator of hash and height (or `CheckPoint`) to solve this.

LLFourn reviewed (changes requested):
This is coming together really nicely. The `sync` code is really easy to follow. This is presents probably the optimal way of using the library if you are creating a service that is meant to be on all the time tracking deposits/withdrawals/LN channels etc.

I've left several questions and ways I think this can be improved.

LLFourn commented:
> The `Emitter::mempool` method is taking ~20seconds (today, with a busy mempool). This is because we need call `getrawtransaction` for every mempool entry that requires re-emission.
> 
> Possible ways to optimize in the future:
> 
>     1. Cache transactions in the emitter.
> 
>     2. Use multiple threads to call `getrawtransaction`.
> 
>     3. A combination of the above.
> 
> 
> This issue should not exist if we sync to tip before emitting from mempool (need to check).

Couldn't we also just have a kind of "stream" of mempool txs that never terminates and just continuously emits mempool txs you haven't seen before? So rename current `Emitter` to `BlockEmitter` and make it only emit blocks. Then create a second thing like `MempoolStream` which polls for new things and emits them when it finds them as best you can with the rpc API (probably via caching). `MempoolStream` would be continuously polled in a separate thread during the lifetime of the application I guess.

Side remark: we may want to name the crate like `bitcoin_rpc_polling` rather than `bitcoin_rpc` given we have other ways of using `bitcoin_rpc` i.e. the wallet which might have different dependencies?

evanlinjin commented:
> Couldn't we also just have a kind of "stream" of mempool txs that never terminates and just continuously emits mempool txs you haven't seen before? So rename current `Emitter` to `BlockEmitter` and make it only emit blocks. Then create a second thing like `MempoolStream` which polls for new things and emits them when it finds them as best you can with the rpc API (probably via caching). `MempoolStream` would be continuously polled in a separate thread during the lifetime of the application I guess.

Currently, `Emitter::mempool` relies on the last block emitted (during the last mempool emission) to determine whether we need to re-emit a mempool tx.

* If the last block emitted is no longer in the best chain, we re-emit everything in the mempool (this can be improved in the future, however, given the rarity of reorgs I think this is good enough).
* If the last block emitted during the last mempool emission is still in the best chain, we only need to emit txs introduced after that given height.

Unless we have some way for the `BlockEmitter` to inform the `MempoolEmitter` about the emitted blocks?

P.S. The re-emission logic works, and does wonders when we are synced close to the tip:

![image](https://github.com/bitcoindevkit/bdk/assets/14065705/ba3efdd3-2e21-48e8-a874-a0eb65cf17d9)

evanlinjin committed:
chain: improvements to `IndexedTxGraph` and `TxGraph` APIs

For `IndexedTxGraph`:
- Remove `InsertTxItem` type (this is too complex).
    - `batch_insert_relevant` now uses a simple tuple `(&tx, anchors)`.
    - `batch_insert` is now also removed, as the same functionality can be
      done elsewhere.
- Add internal helper method `index_tx_graph_changeset` so we don't need
  to create a seprate `TxGraph` update in each method.
- `batch_insert_<relevant>_unconfirmed` no longer takes in an option of
  last_seen.
- `batch_insert_unconfirmed` no longer takes a reference of a
  transaction (since we apply all transactions anyway, so there is no
  need to clone).

For `TxGraph`:
- Add `batch_insert_unconfirmed` method.

LLFourn commented:
> Currently, `Emitter::mempool` relies on the last block emitted (during the last mempool emission) to determine whether we need to re-emit a mempool tx.
> 
>     * If the last block emitted is no longer in the best chain, we re-emit everything in the mempool (this can be improved in the future, however, given the rarity of reorgs I think this is good enough).
> 
>     * If the last block emitted during the last mempool emission is still in the best chain, we only need to emit txs introduced after that given height.

I think I understand the whole solution now thanks. It does seem to be the simplest solution that gives the ordering guarantees. Nice!

evanlinjin committed:
chain: split `IndexedTxGraph::insert_tx` into 3 methods

Instead of inserting anchors and seen_at timestamp in the same method,
we have three separate methods. This makes the API easier to understand
and makes `IndexedTxGraph` more consistent with the `TxGraph` API.

evanlinjin committed:
bitcoind_rpc: rm `BlockHash` from `Emitter::last_mempool_tip`

Instead of comparing the blockhash against the emitted_blocks map
to see whether the block is part of the emitter's best chain, we
reduce the `last_mempool_tip` height to the last agreement height
during the polling logic.

The benefits of this is we have tighter bounds for avoiding re-
emission. Also, it will be easier to replace `emitted_blocks` to
a `CheckPoint` (since we no longer rely on map lookup).

evanlinjin commented:
Currently it takes upwards of 20s to load wallet from db (for the cli example). Probably something to look into for the next release.

danielabrozzoni commented on example-crates/example_cli/src/lib.rs:
```diff
@@ -664,45 +666,213 @@ where
         Commands::ChainSpecific(_) => unreachable!("example code should handle this!"),
         Commands::Address { addr_cmd } => {
             let graph = &mut *graph.lock().unwrap();
-            run_address_cmd(graph, db, network, addr_cmd)
```
In d92709a:

Am I correct in saying that this is just a copy-paste of the internals of `run_address_cmd`? What's the reason behind this? If the method is not needed anymore, can we delete it?

evanlinjin replied:
Woops, duplicates are now removed.

danielabrozzoni commented on example-crates/example_electrum/src/main.rs:
```diff
@@ -66,8 +103,10 @@ type ChangeSet = (
 );
 
 fn main() -> anyhow::Result<()> {
+    // let (args, keymap, index, db, (disk_local_chain, disk_tx_graph)) =
```
In d92709a: these two lines should probably be removed and not commented

danielabrozzoni commented on crates/bitcoind_rpc/src/lib.rs:
```diff
@@ -0,0 +1,243 @@
+//! This crate is used for emitting blockchain data from the `bitcoind` RPC interface (excluding the
```
" from the `bitcoind` RPC interface (excluding the RPC wallet API)."

What do you mean? Is this saying that we don't use any wallet-related RPCs?

evanlinjin replied:
Yes, how do you recommend it worded?

danielabrozzoni replied:
This crate is used for emitting blockchain data from the `bitcoind` RPC interface. It doesn't use the wallet RPC API, so that you can use this crate even with wallet-disabled Bitcoin Core nodes.

How does this sound?

danielabrozzoni commented on crates/bitcoind_rpc/src/lib.rs:
```diff
@@ -0,0 +1,243 @@
+        }
+    }
+
+    /// Emit mempool transactions, alongside their first-seen unix timestamps.
```
Mhh, why are we talking about first-seen timestamp here, while we usually (in the txgraph, from what I've seen) talk about last seen timestamp?

evanlinjin replied:
The RPC API provides the timestamp of when a tx is first seen in the mempool. It would still make sense to prioritize later first-seen timestamps during conflicts.

I think I tried using `seen_at` instead of `last_seen`.

Or we can argue we are still getting the last `first_seen` for a tx coming in from the mempool.

danielabrozzoni commented on crates/bitcoind_rpc/src/lib.rs:
```diff
@@ -0,0 +1,243 @@
+    /// only seen up to block of height `h-1`, we want to re-emit this transaction until the
+    /// receiver has seen the block at height `h`.
+    ///
+    /// In other words, we want to re-emit a transaction if we cannot guarantee it's ancestors are
```
typo: its ancestors

danielabrozzoni commented on crates/bitcoind_rpc/src/lib.rs:
```diff
@@ -0,0 +1,243 @@
+            .get_raw_mempool_verbose()?
+            .into_iter()
+            .filter_map({
+                let latest_time = &mut latest_time;
```
It seems to me that this line was put here just to please the compiler, but actually it works fine even if you remove it. Unless there's some logic difference that I don't see.

evanlinjin replied:
last_seen will be copied into the closure and the original last_seen will stay the same.

danielabrozzoni replied:
> The original last_seen will stay the same.

Is it? The new latest_time is just a mutable reference to the original latest_time, so you're modifying the original as well inside the closure.

But if I'm wrong, and in the closure you have a different latest_time that you're modifying only inside the closure, why is that needed? In the closure you only modify latest_time, you never read it.

danielabrozzoni replied:
Sorry I didn't try, removing the line some tests fail, so it is needed. I just don't get why :(

evanlinjin replied:
latest_time implements copy, so it won't be moved into the closure, it'll get copied.

danielabrozzoni commented on crates/bitcoind_rpc/src/lib.rs:
```diff
@@ -0,0 +1,243 @@
+
+    /// Emit mempool transactions, alongside their first-seen unix timestamps.
+    ///
+    /// Ideally, this method would only emit the same transaction once. However, if the receiver
```
nit: It might be clearer if you put it in a different way, first explaining when a tx might be emitted more than once, and then giving the reason:
```
This method emits each transaction only once, unless when we cannot guarantee the transaction's ancestors
are already emitted.

To understand why, consider a receiver filtering transactions based on...
```

(The first sentence might need some rewording :))

danielabrozzoni commented on crates/bitcoind_rpc/src/lib.rs:
```diff
@@ -0,0 +1,243 @@
+    start_height: u32,
+
+    emitted_blocks: BTreeMap<u32, BlockHash>,
+    last_block: Option<bitcoincore_rpc_json::GetBlockResult>,
```
This needs a bit of documentation to explain when the option is None, as one would expect that it's None only if the Emitter never emitted blocks, but instead you set it to None in various scenarios (no more block, block not in the best chain, etc...).

danielabrozzoni replied:
Same for emitted blocks, we should clarify that it's blocks emitted that are in the best chain, and so we can clear the map if we can't find agreement, or remove some blocks when they get reorged out

danielabrozzoni commented on crates/bitcoind_rpc/tests/test_emitter.rs:
```diff
@@ -0,0 +1,759 @@
+        };
+        let block_hashes = self
+            .client
+            .generate_to_address(count as _, &coinbase_address)?;
```
nit: If generate_to_address takes something different than a usize, why not having mine_blocks take that as well?

danielabrozzoni commented on crates/bitcoind_rpc/tests/test_emitter.rs:
```diff
@@ -0,0 +1,759 @@
+        res
+    }
+
+    fn reorg_empty_blocks(&self, count: usize) -> anyhow::Result<Vec<(usize, BlockHash)>> {
```
nit: reorg_with_empty_blocks might be better, otherwise it seems that you're reorging out empty blocks

danielabrozzoni commented on example-crates/example_bitcoind_rpc_polling/src/main.rs:
```diff
@@ -0,0 +1,366 @@
+const DB_MAGIC: &[u8] = b"bdk_example_rpc";
+const DB_PATH: &str = ".bdk_example_rpc.db";
+
+const CHANNEL_BOUND: usize = 10;
```
nit: needs docs :)

evanlinjin committed:
bitcoind_rpc!: bring back `CheckPoint`s to `Emitter`

* `bdk_chain` dependency is added. In the future, we will introduce a
  separate `bdk_core` crate to contain shared types.
* replace `Emitter::new` with `from_height` and `from_checkpoint`
  * `from_height` emits from the given start height
  * `from_checkpoint` uses the provided cp to find agreement point
* introduce logic that ensures emitted blocks can connect with
  receiver's `LocalChain`
* in our rpc example, we can now `expect()` chain updates to always
  since we are using checkpoints and receiving blocks in order

notmandatory commented on crates/bitcoind_rpc/src/lib.rs:
```diff
@@ -209,7 +210,18 @@ where
+                // The tip during the last mempool emission needs to in the best chain, we reduce
+                // it if it is not.
+                if let Some(h) = emitter.last_mempool_tip.as_mut() {
+                    if dbg!(*h) > dbg!(agreement_h) {
```
Did you mean to leave these `dbg!` macros here?

evanlinjin replied:
removed

evanlinjin committed:
example_bitcoind_rpc: tweaks

* avoid holding mutex lock over io
* document `CHANNEL_BOUND` const
* use the `relevant` variant of `batch_insert_unconfirmed`
* print elapsed time in stdout for various updates

evanlinjin committed:
docs(bitcoind_rpc): better `Emitter::mempool` explanation

Also better docs for `Emitter` fields.

evanlinjin commented:
@danielabrozzoni Sorry, I'm going to ignore your nits due to how long this PR is taking. We can revise `TestEnv` in the future. 

I also think we should put `TestEnv` in a separate crate so it can be shared for testing chain source crates.

LLFourn commented on crates/chain/src/indexed_tx_graph.rs:
```diff
@@ -72,97 +72,197 @@ impl<A: Anchor, I: Indexer> IndexedTxGraph<A, I>
-    /// `anchors` can be provided to anchor the transactions to blocks. `seen_at` is a unix
-    /// timestamp of when the transactions are last seen.
-    pub fn insert_relevant_txs<'t>(
+    /// Items of `txs` are tuples containing the transaction and an optional *last seen* timestamp.
```
```suggestion
    /// Items of `txs` are tuples containing the transaction and an *last seen* timestamp.
```

LLFourn commented on crates/chain/src/indexed_tx_graph.rs:
```diff
@@ -72,97 +72,197 @@ impl<A: Anchor, I: Indexer> IndexedTxGraph<A, I>
+
+    /// Batch insert unconfirmed transactions.
+    ///
+    /// Items of `txs` are tuples containing the transaction and an optional *last seen* timestamp.
```
```suggestion
    /// Items of `txs` are tuples containing the transaction and an *last seen* timestamp.
```

LLFourn reviewed (approved):
Thanks this looks excellent. It would be nice to know why it's taking 20s before merging though? Is it a function of the size of the db file -- how big is it?

vladimirfomene commented on crates/bitcoind_rpc/src/lib.rs:
```diff
@@ -0,0 +1,301 @@
+
+    /// The latest first-seen epoch of emitted mempool transactions. This is used to determine
+    /// whether a mempool transaction is already emitted.
+    last_mempool_time: usize,
```
Why define this as `usize`? Can we have it as `u64` to avoid conversion later?

vladimirfomene commented on crates/bitcoind_rpc/src/lib.rs:
```diff
@@ -0,0 +1,301 @@
+    /// that the block is no longer in the best chain, it will be popped off from here.
+    last_cp: Option<CheckPoint>,
+
+    /// The block result returned from rpc of the last-emitted block. As this result contains the
```
I think using `GetBlockResult` will make this sentence clearer.

evanlinjin commented:
@LLFourn the file is 104MB, with a start height of 778,000

**UPDATE:** Okay I accidentally called `batch_insert_unconfirmed` instead of `batch_insert_relevant_confirmed`. So most of the content is mempool txs.

**UPDATE 2:** Yeah with these settings, when a wallet is fully synced, loading from db takes less than 100ms.

vladimirfomene commented on crates/chain/src/indexed_tx_graph.rs:
```diff
@@ -72,97 +72,197 @@ impl<A: Anchor, I: Indexer> IndexedTxGraph<A, I>
+    /// To filter out irrelevant transactions, use [`batch_insert_relevant_unconfirmed`] instead.
+    ///
+    /// [`batch_insert_relevant_unconfirmed`]: IndexedTxGraph::batch_insert_relevant_unconfirmed
+    pub fn batch_insert_unconfirmed(
```
When will this be useful? Can we add docs around when a user will want to use this API?

evanlinjin replied:
When we want to track transactions, but don't want to filter them out.

vladimirfomene commented on example-crates/example_bitcoind_rpc_polling/src/main.rs:
```diff
@@ -0,0 +1,368 @@
+const STDOUT_PRINT_DELAY: Duration = Duration::from_secs(6);
+/// Delay between mempool emissions.
+const MEMPOOL_EMIT_DELAY: Duration = Duration::from_secs(30);
+/// Delay for commiting to persistance.
```
Nit: committing

vladimirfomene commented on example-crates/example_bitcoind_rpc_polling/src/main.rs:
```diff
@@ -0,0 +1,368 @@
+            }
+
+            let mempool_txs = emitter.mempool()?;
+            let graph_changeset = graph.lock().unwrap().batch_insert_unconfirmed(mempool_txs);
```
Why not use the relevant variant of the `batch_insert_unconfirmed` here?

evanlinjin replied:
Great spot, yes this is a bug.

vladimirfomene reviewed (commented):
Great work @evanlinjin ! Just a couple of questions and nits.

notmandatory reviewed (approved):
Re ACK 85c62532a55cfc94eade4d20ca3075dc4cd4882e

I don't see any critical issues remaining and any final nits can be fixed in follow-up PRs.
